### ISSUE TITLE
Decephalopoded Chaindata

### ISSUE THREAD

**Issue by alecdwm** [TalismanSociety/talisman#1076]
##### Latest Summary before we can merge:

- [x] Balance modules TypeRegistry -> subshape
- [x] Make new tokens stuff prod ready
  - Check that chain and token icons are correct
  - Need to fetch tokens list and miniMetadatas from the new chaindata cache
  - MiniMetadataUpdater needs to be converted from "working" to "reliable" and fast

---

- [ ] Throw some errors deep in the code and find out where they surface, to make sure that we're handling them appropriately
- [x] Update balance module docs

---

- [ ] (Alec/Kheops) make sure that token toggles are working

##### Wallet repo:
- [x] Consolidate balance module libs
- [ ] Consolidate other libs
- [x] Refactor balance modules to use `@talismn/scale` instead of `@talismn/mutate-metadata`
- [x] Refactor `@talismn/chaindata-provider` to pull from github pages endpoint
- [x] Add module (probably internal to `@talismn/balances`) which stores and updates miniMetadatas
  - To start don't bother putting it in a web worker, come back to that (but when designing it keep in mind that it should eventually go in a web worker)
- [x] Update hooks to use module (either run in react useEffect or in spawned web worker)
- [ ] Write new wallet balances integration (balances.store.ts) to use spawned web worker
- [ ] Use balances-react hooks to retrieve balances directly from store
- [ ] ?
- [x] Run metadata stuff in chaindata so that the wallet can just fetch latest status from there (and cache it as a JSON along with chains.json and evm-networks.json) and will only need to run the miniMetadataUpdater on custom / out of date networks
  - [x] This process should use the chainsExtraCache so that it doesn't do unnecessary work
- [ ] ?
- [ ] Update docs & READMEs

##### Chaindata repo:
- [x] Add an action which reads the config files from the repo, derives up to date values for each of the queryable chain properties and then publishes the results to a github pages branch
  - While building, consider whether the action should update every chain at once or if it should update individual chains/batches of chains at a time
- [x] Add a trigger to run the action on (a) a timer, and (b) on changes to the main branch
- [x] Remove `tokens.json` and merge into the `balanceModuleConfigs` chain key
  - Perhaps also rename the `balanceModuleConfigs` key while we're at it

#### Notes/Scope

- Consolidate all balance libs into just `@talismn/balances` and `@talismn/balances-react`
- Consolidate `@talismn/chaindata-provider-extension` into `@talismn/chaindata-provider`
- Consolidate `@talismn/connection-meta` into `@talismn/chain-connector`
- Consolidate chaindata-provider into balances
- Decommission `@talismn/mutate-metadata` (replaced by `@talismn/scale`)

<div />

- Balance libs need a function which using `@talismn/scale` processes `full metadata` and spits out `minimal metadatas` (`miniMetadatas`?), as well as any custom derived variables (e.g. `nominationPoolsPalletId`)
  - It needs to work in the browser, so we cannot use the existing `@talismn/mutate-metadata` library - we'll need to use `@talismn/scale` (based on `subshape`) instead
- A separate function will receive `chaindata chain config`, `user chain config`, `miniMetadatas` and custom derived vars; then it will spit out tokens list

<div />

- Need to schedule metadata updates for chains which have no `miniMetadatas`, or chains with a new `runtimeVersion` (and therefore have out of date `miniMetadatas`)
- Need to include balances lib version/hash as well as `runtimeVersion` so that changing the lib will cause the metadata to be scheduled for a refresh, despite the `runtimeVersion` not upgrading
- Metadata processing ideally will run in a web worker
- Metadata processing needs to be run a queue because there's lots of chains
  - I'm keen on using `@supercharge/promise-pool` to run the queue, also it should be kept in mind that we might not always want the queue to run to completion right **now**, so we need cancellation as well as controls for which out-of-date chains are to be refreshed
  - Also, this needs to support prioritising some chains over others (e.g. chains the user has a balance on)
- Balance fetching needs to take into account whether or not the chain has been processed or not, if not then set balances status to something novel so we can inform the user in the UI (should be similar to but distinct from the `stale` status)

<div />

- Need to move the fields `genesisHash`, `specVersion`, `logo`, etc into a wallet scheduler just like we do with the metadata, as we don't want these to be a github action.

- Need to cache the github json, but this won't include the metadata anymore - so, we will need to also run a script to cache the metadata for all of the chains and bundle it in each wallet release

<div />

- Need to fetch chaindata from github (ideally via a github pages published url for its cdn)
- Need to cache processed data into each release so that users don't need to process every chain immediately after installing/upgrading

<div />

- Triggers for metadata updates:
  - Whenever an extension window is opened, plus:
  - On a once per hour timer for open windows

<div />

- Update balances docs and READMEs to reflect the new functionality
- Include updated instructions for devs who want to use the lib

<div />

- Stretch goal: run miniMetadatas updater module in a web worker
- Stretch goal: add support for user-defined psp22 tokens in wallet settings
- Stretch goal: publish miniMetadatas to the chaindata github pages endpoint
